notification/expression.go

```diff
@@ -30,38 +33,61 @@ type ExpressionRunner struct {
 	CelEnv       map[string]any
 }
 
-func (t ExpressionRunner) logToJobHistory(ctx api.Context, name, errMsg string) {
-	jobHistory := models.NewJobHistory(name, t.ResourceType, t.ResourceID)
-	jobHistory.Start()
-	jobHistory.AddError(errMsg)
-	if err := db.PersistJobHistory(ctx, jobHistory.End()); err != nil {
-		logger.Errorf("error persisting job history: %v", err)
-	}
-}
-
 // Eval evaluates the given expression into a boolean.
 // The expression should return a boolean value that's supported by strconv.ParseBool.
 func (t ExpressionRunner) Eval(ctx api.Context, expression string) (bool, error) {
+	jobHistory := models.NewJobHistory("NotificationFilterEval", t.ResourceType, t.ResourceID).Start()
+	defer func() {
+		passingCurrently := jobHistory.ErrorCount == 0
+
+		if value, found := expressionResultCache.Get(t.ResourceID); found {
+			if passingPreviously, ok := value.(bool); ok {
+				if passingPreviously == passingCurrently {
+					// to avoid excessive db calls, we only save the job history if the expression evaluation changes
+					return
+				}
+			}
+		}
+
+		logs.IfError(db.PersistJobHistory(ctx, jobHistory.End()), "error persisting notification filter evaluation job history")
+		expressionResultCache.SetDefault(t.ResourceID, passingCurrently)
+	}()
+
+	result, err := t.eval(ctx, expression)
+	if err != nil {
+		jobHistory.AddError(err.Error())
```

Comment:
I think this would still create too many events. 

If I have 1 notifications and 10 notifications per minute, it will insert 10 job history errors per minute. 

Why don;t we just put a status and error on the notification itself, if we have an error then we update the table and skip processing the notification entirely, on change of the notification filter we reset the status and error. 

Would probably need a notify to listen to changes on notifications